src/test/java/org/opensearch/knn/index/query/KNNQueryBuilderTests.java

```diff
@@ -181,26 +199,82 @@ public void testDoToQuery_FromModel() {
     }
 
     public void testDoToQuery_InvalidFieldType() throws IOException {
         float[] queryVector = { 1.0f, 2.0f, 3.0f, 4.0f };
-        KNNQueryBuilder knnQueryBuilder = new KNNQueryBuilder("mynumber", queryVector, 1);
+        KNNQueryBuilder knnQueryBuilder = new KNNQueryBuilder("mynumber", queryVector, K);
         Index dummyIndex = new Index("dummy", "dummy");
         QueryShardContext mockQueryShardContext = mock(QueryShardContext.class);
         NumberFieldMapper.NumberFieldType mockNumberField = mock(NumberFieldMapper.NumberFieldType.class);
         when(mockQueryShardContext.index()).thenReturn(dummyIndex);
         when(mockQueryShardContext.fieldMapper(anyString())).thenReturn(mockNumberField);
         expectThrows(IllegalArgumentException.class, () -> knnQueryBuilder.doToQuery(mockQueryShardContext));
     }
+
+    public void testSerialization() throws Exception {
+
+        final KNNQueryBuilder knnQueryBuilder = new KNNQueryBuilder(FIELD_NAME, QUERY_VECTOR, K);
+
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.setVersion(Version.CURRENT);
+            output.writeNamedWriteable(knnQueryBuilder);
+
+            try (StreamInput in = new NamedWriteableAwareStreamInput(output.bytes().streamInput(), writableRegistry())) {
+                in.setVersion(Version.CURRENT);
+                final QueryBuilder deserializedQuery = in.readNamedWriteable(QueryBuilder.class);
+                assertSerialization(deserializedQuery, true);
+            }
+        }
+
+        final KNNQueryBuilder knnQueryBuilderWithFilter = new KNNQueryBuilder(FIELD_NAME, QUERY_VECTOR, K, TERM_QUERY);
+
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.setVersion(Version.CURRENT);
+            output.writeNamedWriteable(knnQueryBuilderWithFilter);
+
+            try (StreamInput in = new NamedWriteableAwareStreamInput(output.bytes().streamInput(), writableRegistry())) {
+                in.setVersion(Version.CURRENT);
+                final QueryBuilder deserializedQuery = in.readNamedWriteable(QueryBuilder.class);
+                assertSerialization(deserializedQuery, false);
+            }
+        }
+
+        // test serialization from < 2.4 versions
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.setVersion(Version.V_2_3_0);
+            output.writeNamedWriteable(knnQueryBuilderWithFilter);
+
+            try (StreamInput in = new NamedWriteableAwareStreamInput(output.bytes().streamInput(), writableRegistry())) {
+                in.setVersion(Version.V_2_3_0);
+                final QueryBuilder deserializedQuery = in.readNamedWriteable(QueryBuilder.class);
+                assertSerialization(deserializedQuery, true);
+            }
+        }
+    }
+
+    private void assertSerialization(final QueryBuilder deserializedQuery, boolean assertFilterIsNull) {
```

Comment:
TERM_QUERY is decided by KNNQueryBuilder but this method does not get the value from the KNNQueryBuilder. 
Also I see there is a room for further optimization.
```
public void testSerialization() throws Exception {
    testSerialization(Version.CURRENT, null);
    testSerialization(Version.CURRENT, TERM_QUERY);
    testSerialization(Version.V_2_3_0, null);
}
```

Reply:
ack